docs/dynamic-bundling/index.md

```diff
@@ -0,0 +1,41 @@
+# Dynamic Bundling
+
+One way to address the gap in serving primitives is a technique we’ll call “dynamic bundling” here.
+Traditional asset deployments require that there’s a strict 1-to-1 mapping between output files generated at build time and network requests at runtime.
+The output files of the build process are served by a simple file server without any additional transformations beyond compression.
+
+Dynamic bundling adds a runtime component that allows for assembly of network responses from file fragments (“chunks”) generated at build time.
+This makes using a simple file server impossible but offers the following advantages:
+
+* Reduced network requests caused by code splitting.
+  Very fine-grained splitting strategies can be used without the overhead and latency anomalies that come with tens or hundreds of individual requests.
+* Reduced number of output files (and build runtime) when generating many variants.
+  E.g. each use of a feature flag (or experiment) can be isolated into one small chunk.
+  The correct code fragment can then be dynamically selected while assembling the bundle.
+  For `n` feature flags, we only need `O(n)` files.
+    - Build-time assembly, on the other hand, is forced to combine files that use many different feature flags,
+      which means generating `O(2n)` output files for `n` feature flags.
+* The content of files can be templated, e.g. to inline minor variants without duplicating big chunks of code in the outputs.
```

Comment:
**Disclaimer:** Controversial feature even among people using it.

I'm not sure what the canonical example is but a simplified example could be:

```scss
/* colors.scss */
$base-color = green;

/* colors-experiment.scss */
$base-color = red;

/* app.scss */
@import 'colors.scss';
@import 'colors-experiment.scss'; // @if(EXPERIMENT_RED_COLORS)

body { background: $base-color; /* many more styles */ }
/* many more rules, also using $base-color */
```

There's two ways to compile this:

1. Try to safely reorder the rules so that everything using `$base-color` can appear in a single block. Somehow ensure that the dependency edge is in just the right spot so it works out.
2. Preserve the variable and resolve it while generating the asset response. This is the kind of strategy this bullet point refers to.

The result of the second approach are templates that may look like this:

```
{{! Generated from colors.scss }}
{{#def}}base-color green{{/def}}

{{! Generated from colors-experiment.scss, included conditionally }}
{{#def}}base-color green{{/def}}

{{! Generated from app.scss }}
body { background: {{base-color}}; /* many more styles */ }
/* many more rules, also using {{base-color}} */
```

Reply:
(Updated the comment to more clearly reflect why "duplicating content" is a likely alternative.)